### PR TITLE
support full storage dump

### PR DESCRIPTION
**Pull request by kopiczko**
This is about being able to retrieve all keys by invoking `storage.List(ctx, "/")`. I need that for migration.

Towards https://github.com/giantswarm/giantswarm/issues/1650

### CODE REVIEW

File: storagetest/storagetest.go [hunk 1]
```diff
@@ -266,3 +303,11 @@ func validKeyVariations(key string) []string {
...
 	}
 }
+
+func sanitize(key string) string {
+	k, err := microstorage.SanitizeKey(key)
+	if err != nil {
+		panic(err)
+	}
+	return k
+}
```
Comment (kopiczko) [inline storagetest/storagetest.go]: This is to make tests more readable.

File: memory/storage.go [hunk 2]
```diff
@@ -104,14 +104,23 @@ func (s *Storage) Exists(ctx context.Context, key string) (bool, error) {
...
 	defer s.mutex.Unlock()
 
+	// Special case.
+	if key == "/" {
+		var list []string
+		for k, _ := range s.data {
+			list = append(list, k)
+		}
+		return list, nil
+	}
```
Comment (kopiczko) [inline memory/storage.go]: memory storage changed a little. the same will affect etcdstorage

File: migration/migration.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package migration
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/microstorage"
+)
+
+func Migrate(ctx context.Context, src, dest microstorage.Storage) error {
```
Comment (xh3b4sd) [inline migration/migration.go]: The order in golang I am used to is usually `dst` before `src`. See https://golang.org/pkg/builtin/#copy for instance.

File: migration/migration.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+		}
+
+		err = dest.Put(ctx, key, v)
+		if err != nil {
+			return microerror.Maskf(err, "putting key=%s value to destination storage", key)
+		}
+	}
+
+	return nil
+}
```
Comment (xh3b4sd) [inline migration/migration.go]: Migrate deserves a backoff somehow.

File: migration/migration.go (hunk 3 above)
Comment (kopiczko) [inline migration/migration.go]: This is even better. I'll change.

File: migration/migration.go (hunk 4 above)
Comment (kopiczko) [inline migration/migration.go]: Actually this sneaked out by mistake. I wanted to extend the test suit and memory impl within this pr. I'll remove migration and create a separate PR.

**Pull request merged**